src/common/proxy_wasm/ngx_proxy_wasm_properties.c

```diff
@@ -911,16 +918,19 @@ ngx_proxy_wasm_properties_set_host(ngx_proxy_wasm_ctx_t *pwctx,
     hpn = (host_props_node_t *)
               ngx_str_rbtree_lookup(&pwctx->host_props_tree, path, hash);
 
-    if (value->data == NULL) {
-        if (hpn) {
-            ngx_rbtree_delete(&pwctx->host_props_tree, &hpn->sn.node);
+    if (hpn) {
+        if (hpn->is_const) {
+            return NGX_DECLINED;
         }
 
-        return NGX_OK;
-    }
-
-    if (hpn) {
         ngx_pfree(pwctx->pool, hpn->value.data);
+
+        if (value->data == NULL && !is_const) {
```

Comment:
It is, because:
* for non-const, setting a `NULL` value means removing an entry from the rbtree;
* for const, setting a `NULL` value means storing a negative-cache entry, so that further gets to the key don't hit the Lua getter
 
That is, preserving the "const cache" behavior consistently for all values returned by the Lua setter, including `return true, null, true`  meaning "success, value is `null`, cache this value".

Reply:
Ok I see, thanks!